README.md
# xtream jupyter template

A simple template for jupyter notebooks.

The extension sets up any new Jupyter Notebook 
with a conventional and general-purpose
template to shape Data Science analysis.

The template includes conventional sections,
like *Data Import*, *Processing* and *References*,
as well as code to perform common operations, like
importing and configuring charting libraries. 

Moreover, it prompts for a notebook name. 
Quite annoyingly. But no worries, you can disable
this one.

## Motivations
Jupyter notebooks are awesome tools: they enable fast 
prototyping and ease result sharing. However, due to 
their flexibility, they are prone to be abused. 

In 
order to help Data Scientist keep their notebooks 
clean, a reasonably flexible yet conventional template
may help. Moreover, such a template may be also a 
productivity tool, which speeds up common setup,
such as library import.

## Repository setup
Please follow this steps to be ready to go:
1. Install the virtual environment by running:
    ```shell
    pipenv install
    ```
2. Enable Jupyter extensions by running:
    ```shell
    jupyter contrib nbextensions install
    ```
This is it. You're ready to go!

## Extension setup
Please follow these steps to get the extension up and 
running:
1. Locate your jupyter extensions root folder - if you are
using a virtual environment, it should be in 
*<venv_root>/Lib/site-packages/jupyter_contrib_nbextensions/nbextensions*
2. Copy the folder xtream-template in this repository
into the jupyter extensions root folder
3. Update the extensions by running: 
    ```shell
    jupyter contrib nbextensions install
    ```
This is it. You're ready to go!

## References
Please consider reading the following resources for
a more comprehensive understanding:
- Will Kohersen, [Set Your Jupyter Notebook up Right with this Extension](https://towardsdatascience.com/set-your-jupyter-notebook-up-right-with-this-extension-24921838a332)
- Will Kohersen, [How to Write a Jupyter Notebook Extension](https://towardsdatascience.com/how-to-write-a-jupyter-notebook-extension-a63f9578a38c)
- Will Kohersen, [Setup extension](https://github.com/WillKoehrsen/Data-Analysis/tree/master/setup)

CHANGELOG
# Changelog


## (unreleased)

### New

* Make changelog readable (0.1.0) [Emanuele Fabbiani]

* First commit. [Emanuele Fabbiani]

  Introduce template based on github code (see readme).

### Other

* Add CHANGELOG. [Emanuele Fabbiani]

* Initial commit. [Emanuele Fabbiani]




xtream-template/main.js
define(['base/js/namespace', 'base/js/events'], function (Jupyter, events) {
    // Template cells including markdown and imports
    var setUp = function () {
        Jupyter.notebook.insert_cell_at_index('markdown', 0)
            .set_text(`# Introduction
State notebook purpose here`)
        Jupyter.notebook.insert_cell_at_index('markdown', 1).set_text(`### Imports
Import libraries and write settings here.`)
        // Define imports and settings
        Jupyter.notebook.insert_cell_at_index('code', 2)
            .set_text(`# Data manipulation
import pandas as pd
import numpy as np

# Options for pandas
pd.options.display.max_columns = 50
pd.options.display.max_rows = 30

# Display all cell outputs
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

from IPython import get_ipython
ipython = get_ipython()

# autoreload extension
if 'autoreload' not in ipython.extension_manager.loaded:
    %load_ext autoreload

%autoreload 2

# Visualizations
import plotly.plotly as py
import plotly.graph_objs as go
from plotly.offline import iplot, init_notebook_mode
init_notebook_mode(connected=True)

import cufflinks as cf
cf.go_offline(connected=True)
cf.set_config_file(theme='white')`)
        Jupyter.notebook.insert_cell_at_index('markdown', 3)
            .set_text(`# Analysis/Modeling
Do work here`)
        Jupyter.notebook.insert_cell_at_index('markdown', 4).set_text(`# Results
Show graphs and stats here`)
        Jupyter.notebook.insert_cell_at_index('markdown', 5)
            .set_text(`# Conclusions and Next Steps
Summarize findings here`)
        // Run all cells
        Jupyter.notebook.execute_all_cells()
    }
    // Prompts user to enter name for notebook
    var promptName = function () {
        // Open rename notebook box if 'Untitled' in name
        if (Jupyter.notebook.notebook_name.search('Untitled') != -1) {
            document.getElementsByClassName('filename')[0].click()
        }
    }

    // Run on start
    function load_ipython_extension() {
        // Add default cells for new notebook
        if (Jupyter.notebook.get_cells().length === 1) {
            setTimeout(setUp, 500)
        } else {
            promptName()
        }
    }

    // Run when cell is executed
    events.on('execute.CodeCell', function () {
        promptName()
    })
    // Run when notebook is saved
    events.on('before_save.Notebook', function () {
        promptName()
    })
    return {
        load_ipython_extension: load_ipython_extension
    }
})


xtream-template/xtream_template.js
define(['require', 'jquery', 'base/js/namespace', 'base/js/events'],
    function (requirejs, $, Jupyter, JupyterEvents) {
        "use strict";

        class CellContent {
            constructor(cell_type, content) {
                this.cell_type = cell_type;
                this.content = content;
            }
        }

        const intro_text = new CellContent('markdown',
            `# Title
The title of the notebook should be coherent with file name. Namely, file name should be:    
*author's initials_progressive number_title.ipynb*    
For example:    
*EF_01_Data Exploration.ipynb*

## Purpose
State the purpose of the notebook.

## Methodology
Quickly describe assumptions and processing steps.

## WIP - improvements
Use this section only if the notebook is not final.

Notable TODOs:
- todo 1;
- todo 2;
- todo 3.

## Results
Describe and comment the most important results.

## Suggested next steps
State suggested next steps, based on results obtained in this notebook.`
        );

        const setup_library_text = new CellContent('markdown',
            `# Setup

## Library import
We import all the required Python libraries`
        );

        const setup_library_snippet = new CellContent('code',
            `# Data manipulation
import pandas as pd
import numpy as np

# Options for pandas
pd.options.display.max_columns = 50
pd.options.display.max_rows = 30

# Visualizations
import plotly
import plotly.graph_objs as go
import plotly.offline as ply
plotly.offline.init_notebook_mode(connected=True)

import cufflinks as cf
cf.go_offline(connected=True)
cf.set_config_file(theme='white')

import matplotlib as plt

# Autoreload extension
if 'autoreload' not in get_ipython().extension_manager.loaded:
    %load_ext autoreload
    
%autoreload 2`
        );

        const setup_local_library_text = new CellContent('markdown',
            `## Local library import
We import all the required local libraries libraries`
        );

        const setup_local_library_snippet = new CellContent('code',
            `# Include local library paths
import sys
# sys.path.append('path/to/local/lib') # uncomment and fill to import local libraries

# Import local libraries`
        );

        const parameter_text = new CellContent('markdown',
            `# Parameter definition
We set all relevant parameters for our notebook. By convention, parameters are uppercase, while all the 
other variables follow Python's conventions.`
        );

        const data_import_text = new CellContent('markdown',
            `
# Data import
We retrieve all the required data for the analysis.`
        );

        const data_processing_text = new CellContent('markdown',
            `# Data processing
Put here the core of the notebook. Feel free di further split this section into subsections.`
        );

        const reference_text = new CellContent('markdown',
            `# References
We report here relevant references:
1. author1, article1, journal1, year1, url1
2. author2, article2, journal2, year2, url2`
        );

        // define default values for config parameters
        let params = {
            insert_template_on_creation: true,
            ask_title_change_if_untitled: true
        };

        const add_sections = function () {
            const section_templates = [
                intro_text,
                setup_library_text,
                setup_library_snippet,
                setup_local_library_text,
                setup_local_library_snippet,
                parameter_text,
                data_import_text,
                data_processing_text,
                reference_text
            ];

            section_templates.forEach(function (item, index) {
                Jupyter.notebook.insert_cell_at_index(item.cell_type, index).set_text(item.content);
            });
        };

        // Prompts user to enter name for notebook
        const prompt_name = function () {
            // Open rename notebook box if 'Untitled' in name
            if (Jupyter.notebook.notebook_name.includes('Untitled')) {
                document.getElementsByClassName('filename')[0].click();
            }
        };

        // define action, register with ActionHandler instance
        const prefix = 'auto';
        const action_name = 'inset-template-cells';
        const action = {
            icon: 'fa-file-text-o',
            help: 'Insert template cells at the beginning of the notebook',
            help_index: 'zz',
            id: 'insert_template_cells',
            handler: add_sections
        };

        const initialize = function () {
            // update params with any specified in the server's config file.
            // the "thisextension" value of the Jupyter notebook config's
            // data may be undefined, but that's ok when using JQuery's extend
            $.extend(true, params, Jupyter.notebook.config.data.xtream_template);

            // add our extension's css to the page
            $('<link/>')
                .attr({
                    rel: 'stylesheet',
                    type: 'text/css',
                    href: requirejs.toUrl('./xtream_template.css')
                })
                .appendTo('head');

            // add sections to the notebook
            if (Jupyter.notebook.get_cells().length === 1 && params.insert_template_on_creation) {
                add_sections();
            }

            if (params.ask_title_change_if_untitled) {
                // Run when cell is executed
                JupyterEvents.on('execute.CodeCell', prompt_name);

                // Run when notebook is saved
                JupyterEvents.on('before_save.Notebook', prompt_name);
            }

            // register actions with ActionHandler instance
            let action_full_name = Jupyter.keyboard_manager.actions.register(action, action_name, prefix);

            // create toolbar button
            Jupyter.toolbar.add_buttons_group([action_full_name]);
        };

        // The specially-named function load_ipython_extension will be called
        // by the notebook when the nbextension is to be loaded.
        // It mustn't take too long to execute however, or the notebook will
        // assume an error has occurred.
        const load_ipython_extension = function () {
            // Once the config has been loaded, do everything else.
            // The loaded object is a javascript Promise object, so the then
            // call return immediately. See
            // https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise
            // for details.
            return Jupyter.notebook.config.loaded.then(initialize);
        };

        // return object to export public methods
        return {
            load_ipython_extension: load_ipython_extension
        };
    }
);
